src/cursor/cur_file.c

```diff
@@ -75,6 +75,76 @@ __curfile_equals(WT_CURSOR *a, WT_CURSOR *b, int *equalp)
     API_END_RET(session, ret);
 }
 
+/*
+ * __curfile_set_key --
+ *     WT_CURSOR->set_key implementation for file cursor.
+ */
+static void
+__curfile_set_key(WT_CURSOR *cursor, ...)
+{
+    va_list ap;
+
+    va_start(ap, cursor);
+    F_CLR((WT_CURSOR_BTREE *)cursor, WT_CBT_RESEARCH);
+    WT_IGNORE_RET(__wt_cursor_set_keyv(cursor, cursor->flags, ap));
+    va_end(ap);
+}
+
+/*
+ * __curfile_research --
+ *     Research the saved key to regain the cursor position
+ */
+static int
+__curfile_research(WT_CURSOR *cursor)
+{
+    WT_CURSOR_BTREE *cbt;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    cbt = (WT_CURSOR_BTREE *)cursor;
+    session = CUR2S(cursor);
+
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_EXT))
+        return (__wt_panic(session, EINVAL, "Research flag is set without a search key"));
+
+    ret = __wt_btcur_search(cbt);
+
+    if (ret != 0)
+        return (
+          __wt_panic(session, ret, "Failed to research to regain the saved cursor key position"));
+
+    /* Search maintains a position, key and value. */
+    WT_ASSERT(session,
+      F_ISSET(cbt, WT_CBT_ACTIVE) && F_MASK(cursor, WT_CURSTD_KEY_SET) == WT_CURSTD_KEY_INT &&
+        F_MASK(cursor, WT_CURSTD_VALUE_SET) == WT_CURSTD_VALUE_INT);
+    
+    F_CLR(cbt, WT_CBT_RESEARCH);
+
+    return (0);
+}
+
+/*
+ * __curfile_release_page --
+ *     Copy the return key value to the local buffer and release the page.
+ */
+static int
+__curfile_release_page(WT_CURSOR *cursor)
+{
+    WT_CURSOR_BTREE *cbt;
+    WT_SESSION_IMPL *session;
+
+    cbt = (WT_CURSOR_BTREE *)cursor;
+    session = CUR2S(cursor);
+
+    WT_RET(__wt_buf_set(session, &cursor->key, cursor->key.data, cursor->key.size));
+    WT_RET(__wt_buf_set(session, &cursor->value, cursor->value.data, cursor->value.size));
+    WT_RET(__wt_btcur_reset(cbt));
+    F_SET(cursor, WT_CURSTD_KEY_EXT | WT_CURSTD_VALUE_EXT);
+    F_SET(cbt, WT_CBT_RESEARCH);
```

Comment:
I expected this to be a `CURSTD` flag since it's above the btree layer?

Reply:
I think this is a implementation detail at the btree layer.

Reply:
Are you sure? There is another flag in there, `WT_CBT_READ_ONCE`, but otherwise we're reasonably clean about not setting `CBT` flags in the cursor/XXX code.

Reply:
And, in second thoughts: you might be correct, in which case, this could/should be implemented in the btree layer.